### PR TITLE
Changes to codec and preemphasis perturbation

### PR DESCRIPTION
**Pull request by Max-Ryujin**
These are older changes that change the way codec and preemphasis work. 
- For non-perturbed segments the default value is applied instead of none.
- mu law randomizes over mu now.
- during recognition, perturbation can be removed from the config and is replaced by additional layers that apply the default value for mu law or preemphasis.

### CODE REVIEW

File: users/vieting/experiments/switchboard/ctc/feat/baseline_args.py [hunk 1]
```diff
@@ -124,14 +124,28 @@ def get_nn_args_single(
...
+    preemphasis_perturbation = False
+    codec_perturbation = False
+    # If audio perturbation is enabled, we need to add the perturbation layers to the feature network
+    extra_args = recog_args.get("extra_args", {})
+    audio_perturb_args = extra_args.get("audio_perturb_args", {})
+    preemphasis_perturbation = "preemphasis" in audio_perturb_args
+    codec_perturbation = "codecs" in audio_perturb_args
+    extra_args.pop("audio_perturb_args", None)
+    extra_args.pop("audio_perturb_runner", None)
+
```
Comment (vieting) [inline users/vieting/experiments/switchboard/ctc/feat/baseline_args.py:138]: I think it would be better to do all of this inside `get_returnn_config()`.

File: users/vieting/experiments/switchboard/ctc/feat/baseline_args.py [hunk 2]
```diff
@@ -237,6 +253,40 @@ def get_returnn_config(
         # network["source"] = specaug_layer_jingjing(in_layer=["features"])
         pass
 
+    if audio_perturbation and recognition:
+        if preemphasis_perturbation:
+            for layer in feature_net["subnetwork"]:
+                layer_config = feature_net["subnetwork"][layer]
+                if layer_config.get("class") != "variable":
+                    if feature_net["subnetwork"][layer].get("from") == "data":
+                        feature_net["subnetwork"][layer]["from"] = "preemphasis"
```
Comment (vieting) [inline users/vieting/experiments/switchboard/ctc/feat/baseline_args.py:262]: Was the check for `"variable"` just so `.get("from")` would not crash?
```suggestion
                if feature_net["subnetwork"][layer].get("from", None) == "data":
                    feature_net["subnetwork"][layer]["from"] = "preemphasis"
```

File: users/vieting/experiments/switchboard/ctc/feat/baseline_args.py [hunk 3]
```diff
@@ -237,6 +253,40 @@ def get_returnn_config(
...
+            feature_net["subnetwork"]["preemphasis"] = PreemphasisNetwork(alpha=0.97).get_as_subnetwork()
```
Comment (vieting) [inline users/vieting/experiments/switchboard/ctc/feat/baseline_args.py:263]: We should maybe introduce a default value in the perturbation params and use it here as well.

File: users/vieting/experiments/switchboard/ctc/feat/baseline_args.py [hunk 4]
```diff
@@ -237,6 +253,40 @@ def get_returnn_config(
...
+        elif codec_perturbation:
+            for layer in feature_net["subnetwork"]:
+                layer_config = feature_net["subnetwork"][layer]
+                if layer_config.get("class") != "variable":
+                    if feature_net["subnetwork"][layer].get("from") == "data":
+                        feature_net["subnetwork"][layer]["from"] = "codec"
```
Comment (vieting) [inline users/vieting/experiments/switchboard/ctc/feat/baseline_args.py:269]:
```suggestion
                if feature_net["subnetwork"][layer].get("from", None) == "data":
                    feature_net["subnetwork"][layer]["from"] = "codec"
```

File: users/vieting/experiments/switchboard/ctc/feat/baseline_args.py [hunk 5]
```diff
@@ -237,6 +253,40 @@ def get_returnn_config(
...
+            feature_net["subnetwork"]["codec"] = {
+                "class": "subnetwork",
+                "from": ["data"],
+                "subnetwork": {
+                    "safe_input": {"class": "eval", "from": "data", "eval": "tf.clip_by_value(source(0), -1.0, 1.0)"},
```
Comment (vieting) [inline users/vieting/experiments/switchboard/ctc/feat/baseline_args.py:274]: Why is this needed?

**Review by vieting — CHANGES_REQUESTED** [review]

File: users/vieting/experiments/switchboard/ctc/feat/network_helpers/perturbation.py [hunk 6]
```diff
@@ -117,25 +119,28 @@ def preemphasis_numpy(waveform, coeff=0.97):
 
         if random_state.random() < factor.prob:
             preemphasis_coefficient = random_state.random() * (factor.max - factor.min) + factor.min
-            audio = preemphasis_numpy(audio, coeff=preemphasis_coefficient)
-
-        return audio
+            return preemphasis_numpy(audio, coeff=preemphasis_coefficient)
+        else:
+            return preemphasis_numpy(audio)
```
[the remark this answers is not in the document]
Comment (Max-Ryujin) [reply users/vieting/experiments/switchboard/ctc/feat/network_helpers/perturbation.py:124]: I will rerun premphasis experiments anyway.

File: users/vieting/experiments/switchboard/ctc/feat/baseline_args.py (hunk 5 above)
Comment (Max-Ryujin) [reply users/vieting/experiments/switchboard/ctc/feat/baseline_args.py:275]: To ensure the data are in the correct range for the mu law codec.

File: users/vieting/experiments/switchboard/ctc/feat/network_helpers/perturbation.py [hunk 7]
```diff
@@ -107,35 +109,40 @@ def sox(self, audio, sample_rate, random_state):
...
     def apply_codecs(audio, sample_rate, random_state, codecs):
+        import numpy as np
+
         for codec in codecs:
-            prob = codec.pop("prob", 1.0)
-            if random_state.random() < prob:
-                if codec.get("encoding") == "ULAW":
-                    # check if audio is in the right range for mu-law encoding
-                    if np.max(np.abs(audio)) > 1.0:
-                        raise ValueError("Audio must be in the range [-1, 1] for mu-law encoding.")
```
Comment (vieting) [inline users/vieting/experiments/switchboard/ctc/feat/network_helpers/perturbation.py:132]: Why is this check removed?

**Comment by Max-Ryujin** [comment]
I changed the args in `experiments.py` as well to be in line with the new args structure, but I want to note that the `experiments.py` needs significant changes later anyway and does not run without additional changes to `/i6_experiments/users/vieting/experiments/switchboard/transducer/feat/data.py`

File: users/vieting/experiments/switchboard/ctc/feat/network_helpers/perturbation.py (hunk 7 above)
Comment (Max-Ryujin) [reply users/vieting/experiments/switchboard/ctc/feat/network_helpers/perturbation.py:132]: Not sure. I added it again

File: users/vieting/experiments/switchboard/ctc/feat/baseline_args.py (hunk 5 above)
Comment (vieting) [reply users/vieting/experiments/switchboard/ctc/feat/baseline_args.py:269]: Oh I see, there used to be an assertion in the perturbation code which was removed. See my comment there.

I think an assertion is much better suited. With this case, you'll not notice if the data is not in the correct range and just clip it which could seriously hamper the performance. Can you do an assertion here as well?

**Comment by vieting** [comment]
> I changed the args in `experiments.py` as well to be in line with the new args structure, but I want to note that the `experiments.py` needs significant changes later anyway and does not run without additional changes to `/i6_experiments/users/vieting/experiments/switchboard/transducer/feat/data.py`

Why is this related to the transducer data? And why doesn't it run like this?

File: users/vieting/experiments/switchboard/ctc/feat/baseline_args.py (hunk 5 above)
Comment (Max-Ryujin) [reply users/vieting/experiments/switchboard/ctc/feat/baseline_args.py:269]: I dont think I can. There is no way for me to check the values of the data before I configure the network and later it would be catched by the readded check in perturbation. I will just remove the layer

**Comment by Max-Ryujin** [comment]
ctc experiments use the same `data.py` to get the trainingsdata. 
`get_returnn_ogg_datasets `is missing the pre_process argument that is needed for audio perturbation.
I will add this in a later PR as it is not related to the changes to preemphasis and codec perturbation

File: users/vieting/experiments/switchboard/ctc/feat/baseline_args.py (hunk 5 above)
Comment (vieting) [reply users/vieting/experiments/switchboard/ctc/feat/baseline_args.py]: No the check is not active during recognition. Can't you somehow assert this in tf in that layer?

File: users/vieting/experiments/switchboard/ctc/feat/baseline_args.py [hunk 8]
```diff
@@ -237,6 +240,47 @@ def get_returnn_config(
...
+        extra_args.pop("audio_perturb_runner", None)
+
+    if audio_perturbation and recognition:
+        if preemphasis_perturbation:
+            for layer in feature_net["subnetwork"]:
+                if feature_net["subnetwork"][layer].get("from", None) == "data":
+                    feature_net["subnetwork"][layer]["from"] = "preemphasis"
+            audio_perturb_args = extra_args.get("audio_perturb_args", {})
+            default_coeff = audio_perturb_args.get("default_coeff", 0.97)
+            feature_net["subnetwork"]["preemphasis"] = PreemphasisNetwork(alpha=default_coeff).get_as_subnetwork()
```
Comment (vieting) [inline users/vieting/experiments/switchboard/ctc/feat/baseline_args.py:260]: I'd prefer not to have a hidden default here. Rather make it explicit. If it doesn't exist, it should crash.
```suggestion
            alpha = audio_perturb_args.get("default")
            feature_net["subnetwork"]["preemphasis"] = PreemphasisNetwork(alpha=alpha).get_as_subnetwork()
```

File: users/vieting/experiments/switchboard/ctc/feat/baseline_args.py [hunk 9]
```diff
@@ -237,6 +240,47 @@ def get_returnn_config(
...
+                    },
+                    "output": {
+                        "class": "eval",
+                        "from": "magnitude",
+                        "eval": "tf.sign(source(0)) * source(1)",
+                        "from": ["data", "magnitude"],
+                    },
+                },
+                "trainable": False,
+            }
```
Comment (vieting) [inline users/vieting/experiments/switchboard/ctc/feat/baseline_args.py]: You can just put that in a single `EvalLayer` instead of a subnetwork. I think that's easier to read.

File: users/vieting/experiments/switchboard/ctc/feat/experiments.py [hunk 10]
```diff
@@ -319,8 +319,8 @@ def run_scf_audio_perturbation():
         {"tempo": {"prob": 0.6, "minimum": 0.83, "maximum": 1.17}},
         {"tempo": {"prob": 0.6, "minimum": 0.9, "maximum": 1.1}},
         {"tempo": {"prob": 0.6, "minimum": 0.8, "maximum": 1.2}},
-        {"preemphasis": {"prob": 0.9, "minimum": 0.9, "maximum": 1.0}},
-        {"preemphasis": {"prob": 1.0, "minimum": 1.0, "maximum": 1.0}},
+        {"preemphasis": {"prob": 0.9, "minimum": 0.9, "maximum": 1.0, "default_coeff": 0.97}},
+        {"preemphasis": {"prob": 1.0, "minimum": 1.0, "maximum": 1.0, "default_coeff": 0.97}},
```
Comment (vieting) [inline users/vieting/experiments/switchboard/ctc/feat/experiments.py:323]:
```suggestion
        {"preemphasis": {"prob": 0.9, "minimum": 0.9, "maximum": 1.0, "default": 0.97}},
        {"preemphasis": {"prob": 1.0, "minimum": 1.0, "maximum": 1.0, "default": 0.97}},
```

File: users/vieting/experiments/switchboard/ctc/feat/network_helpers/perturbation.py [hunk 11]
```diff
@@ -49,6 +49,8 @@ def __init__(
             - 'format' (str): The audio format such as 'wav', 'vorbis' etc.
             - 'encoding' or 'compression' (str/float): The encoding or compression technique and its level to be used.
             - 'prob' (float): The probability of applying this specific codec.
+            - 'minimum' (int): Minimum value for codec if needed. E.g., default for mu is 255.
+            - 'maximum' (int): Maximum value for codec if needed.
             Example: [{"format": "wav", "encoding": "ULAW", "prob": 0.4}]
 
         :param preemphasis: A dictionary containing parameters for the preemphasis filter.
```
Comment (vieting) [inline users/vieting/experiments/switchboard/ctc/feat/network_helpers/perturbation.py:56]: Please add `"default"` to docstring.

File: users/vieting/experiments/switchboard/ctc/feat/network_helpers/perturbation.py [hunk 12]
```diff
@@ -69,7 +71,7 @@ def __init__(
         self._perturbations = []
         self.non_linearity = non_linearity
         if preemphasis:
-            self._perturbations.append(functools.partial(self.preemphasis, factor=PerturbationFactor(**preemphasis)))
+            self._perturbations.append(functools.partial(self.preemphasis, preemphasis_args=preemphasis))
```
Comment (vieting) [inline users/vieting/experiments/switchboard/ctc/feat/network_helpers/perturbation.py:74]: You can extend `PerturbationFactor` with a default value, that would be cleaner than the current solution.

File: users/vieting/experiments/switchboard/ctc/feat/network_helpers/perturbation.py [hunk 13]
```diff
@@ -107,35 +109,42 @@ def sox(self, audio, sample_rate, random_state):
...
-                if codec.get("encoding") == "ULAW":
-                    # check if audio is in the right range for mu-law encoding
-                    if np.max(np.abs(audio)) > 1.0:
-                        raise ValueError("Audio must be in the range [-1, 1] for mu-law encoding.")
-                    # standard value for mu-law encoding
-                    mu = 255.0
-                    # mu-law encoding formula
-                    audio = np.sign(audio) * np.log1p(mu * np.abs(audio)) / np.log1p(mu)
+            if codec.get("encoding") == "ULAW":
+                prob = codec.pop("prob", 1.0)
```
Comment (vieting) [inline users/vieting/experiments/switchboard/ctc/feat/network_helpers/perturbation.py:133]: Can't you also here use a `PerturbationFactor`"

**Review by vieting — CHANGES_REQUESTED** [review]

File: users/vieting/experiments/switchboard/ctc/feat/baseline_args.py (hunk 5 above)
Comment (Max-Ryujin) [reply users/vieting/experiments/switchboard/ctc/feat/baseline_args.py]: Done.

File: users/vieting/experiments/switchboard/ctc/feat/baseline_args.py [hunk 14]
```diff
@@ -124,14 +124,17 @@ def get_nn_args_single(
         **(returnn_args or {}),
     )
 
+    recog_args = returnn_args.copy() if returnn_args else {}
+    feature_recog_net = copy.deepcopy(feature_net)
+
```
[the remark this answers is not in the document]
Comment (Max-Ryujin) [reply users/vieting/experiments/switchboard/ctc/feat/baseline_args.py:129]: The network for recognition must be separate because I need to potentially modify it

File: users/vieting/experiments/switchboard/ctc/feat/baseline_args.py (hunk 9 above)
Comment (Max-Ryujin) [reply users/vieting/experiments/switchboard/ctc/feat/baseline_args.py]: With the addition of the assert_range layer, I think it can stay a subnetwork.

File: users/vieting/experiments/switchboard/ctc/feat/network_helpers/perturbation.py (hunk 13 above)
Comment (Max-Ryujin) [reply users/vieting/experiments/switchboard/ctc/feat/network_helpers/perturbation.py:134]: See above

File: users/vieting/experiments/switchboard/ctc/feat/network_helpers/perturbation.py (hunk 12 above)
Comment (Max-Ryujin) [reply users/vieting/experiments/switchboard/ctc/feat/network_helpers/perturbation.py:75]: functions that need more than these three get the dict directly.

File: users/vieting/experiments/switchboard/ctc/feat/baseline_args.py (hunk 14 above)
[the remark this answers is not in the document]
Comment (vieting) [reply users/vieting/experiments/switchboard/ctc/feat/baseline_args.py:129]: But why not in `get_returnn_config()` in the `if` branch where you know that you need to modify?

File: users/vieting/experiments/switchboard/ctc/feat/baseline_args.py [hunk 15]
```diff
@@ -237,6 +240,50 @@ def get_returnn_config(
...
 
+    if audio_perturbation and recognition:
+        audio_perturb_args = extra_args.get("audio_perturb_args", {})
+        if "preemphasis" in audio_perturb_args:
+            for layer in feature_net["subnetwork"]:
+                if feature_net["subnetwork"][layer].get("from", None) == "data":
+                    feature_net["subnetwork"][layer]["from"] = "preemphasis"
+            alpha = audio_perturb_args.get("default")
+            feature_net["subnetwork"]["preemphasis"] = PreemphasisNetwork(alpha=alpha).get_as_subnetwork()
+        elif "codecs" in audio_perturb_args:
```
Comment (vieting) [inline users/vieting/experiments/switchboard/ctc/feat/baseline_args.py:251]: Can't this be both?
```suggestion
        else "codecs" in audio_perturb_args:
```

File: users/vieting/experiments/switchboard/ctc/feat/network_helpers/perturbation.py (hunk 12 above)
Comment (vieting) [reply users/vieting/experiments/switchboard/ctc/feat/network_helpers/perturbation.py:75]: But wouldn't it be nicer to add an optional default value?

**Review by vieting — CHANGES_REQUESTED** [review]

File: users/vieting/experiments/switchboard/ctc/feat/baseline_args.py (hunk 14 above)
[the remark this answers is not in the document]
Comment (Max-Ryujin) [reply users/vieting/experiments/switchboard/ctc/feat/baseline_args.py:129]: Done

File: users/vieting/experiments/switchboard/ctc/feat/baseline_args.py [hunk 16]
```diff
@@ -237,6 +237,51 @@ def get_returnn_config(
         # network["source"] = specaug_layer_jingjing(in_layer=["features"])
         pass
 
+    if audio_perturbation and recognition:
+        feature_recog_net = copy.deepcopy(feature_net)
```
Comment (vieting) [inline users/vieting/experiments/switchboard/ctc/feat/baseline_args.py:241]:
```suggestion
        feature_net = copy.deepcopy(feature_net)
```
`feature_recog_net` is not used lateron. Also rename the usages below.

**Review by vieting — CHANGES_REQUESTED** [review]

File: users/vieting/experiments/switchboard/ctc/feat/network_helpers/perturbation.py (hunk 12 above)
Comment (Max-Ryujin) [reply users/vieting/experiments/switchboard/ctc/feat/network_helpers/perturbation.py:75]: I will do this

File: users/vieting/experiments/switchboard/ctc/feat/network_helpers/perturbation.py (hunk 13 above)
Comment (Max-Ryujin) [reply users/vieting/experiments/switchboard/ctc/feat/network_helpers/perturbation.py:133]: Codec is a special case since it is a list. As long as I just use mu law, it would be possible to use PerturbationFactor but to make it possible to add others later, I would keep it that way.

File: users/vieting/experiments/switchboard/ctc/feat/baseline_args.py [hunk 17]
```diff
@@ -237,6 +237,54 @@ def get_returnn_config(
...
 
+    if audio_perturbation and recognition:
+        feature_net = copy.deepcopy(feature_net)
+        audio_perturb_args = extra_args.get("audio_perturb_args", {})
+        source_layer = "data"
+        if "preemphasis" in audio_perturb_args:
+            for layer in feature_net["subnetwork"]:
+                if feature_net["subnetwork"][layer].get("from", None) == source_layer:
+                    feature_net["subnetwork"][layer]["from"] = "preemphasis"
+                    source_layer = "preemphasis"
```
Comment (vieting) [inline users/vieting/experiments/switchboard/ctc/feat/baseline_args.py:248]:
```suggestion
            source_layer = "preemphasis"
```
It has to be done after looping through the layers, right?

File: users/vieting/experiments/switchboard/ctc/feat/baseline_args.py [hunk 18]
```diff
@@ -237,6 +237,54 @@ def get_returnn_config(
...
+            alpha = audio_perturb_args.get("default")
+            feature_net["subnetwork"]["preemphasis"] = PreemphasisNetwork(alpha=alpha).get_as_subnetwork()
+        if "codecs" in audio_perturb_args:
+            for layer in feature_net["subnetwork"]:
+                if feature_net["subnetwork"][layer].get("from", None) == source_layer:
+                    feature_net["subnetwork"][layer]["from"] = "codec"
+                    source_layer = "codec"
```
Comment (vieting) [inline users/vieting/experiments/switchboard/ctc/feat/baseline_args.py:255]: It has to be done after looping through the layers, right?
```suggestion
            source_layer = "codec"
```

File: users/vieting/experiments/switchboard/ctc/feat/network_helpers/perturbation.py [hunk 19]
```diff
@@ -110,32 +114,37 @@ def sox(self, audio, sample_rate, random_state):
...
         import numpy as np
 
-        def preemphasis_numpy(waveform, coeff=0.97):
+        def preemphasis_numpy(waveform, coeff):
             waveform = np.copy(waveform)
             waveform[..., 1:] -= coeff * waveform[..., :-1]
             return waveform
 
-        if random_state.random() < factor.prob:
+        if random_state.random() < factor.get("prob", None):
```
Comment (vieting) [inline users/vieting/experiments/switchboard/ctc/feat/network_helpers/perturbation.py:122]: Why is this needed?

File: users/vieting/experiments/switchboard/ctc/feat/baseline_args.py (hunk 17 above)
[the remark this answers is not in the document]
Comment (Max-Ryujin) [reply users/vieting/experiments/switchboard/ctc/feat/baseline_args.py:248]: Perturbation is done first and preemphasis is done after in the network. If preemphasis is also done as a perturbation than it is done before aswell. I never did codec and preemphasis perturbation together, since both did not look promising. I am not sure what the correct order would be if I would do this

File: users/vieting/experiments/switchboard/ctc/feat/network_helpers/perturbation.py (hunk 19 above)
Comment (Max-Ryujin) [reply users/vieting/experiments/switchboard/ctc/feat/network_helpers/perturbation.py:122]: It is not needed

File: users/vieting/experiments/switchboard/ctc/feat/baseline_args.py [hunk 20]
```diff
@@ -237,6 +237,54 @@ def get_returnn_config(
...
+    if audio_perturbation and recognition:
+        feature_net = copy.deepcopy(feature_net)
+        audio_perturb_args = extra_args.get("audio_perturb_args", {})
+        source_layer = "data"
+        if "preemphasis" in audio_perturb_args:
+            for layer in feature_net["subnetwork"]:
+                if feature_net["subnetwork"][layer].get("from", None) == source_layer:
+                    feature_net["subnetwork"][layer]["from"] = "preemphasis"
+                    source_layer = "preemphasis"
+            alpha = audio_perturb_args.get("default")
```
[the remark this answers is not in the document]
Comment (Max-Ryujin) [reply users/vieting/experiments/switchboard/ctc/feat/baseline_args.py]: Yes

File: users/vieting/experiments/switchboard/ctc/feat/baseline_args.py (hunk 17 above)
[the remark this answers is not in the document]
Comment (Max-Ryujin) [reply users/vieting/experiments/switchboard/ctc/feat/baseline_args.py]: I do not intend to do combination experiments with mu law and preemphasis atm. I could make them mutually exclusive, which would make the code simpler and avoid this problem.

[the remark this answers is not in the document]
Comment (vieting) [reply users/vieting/experiments/switchboard/ctc/feat/baseline_args.py]: Yes, please do that. Otherwise, it can lead to hidden errors.

File: users/vieting/experiments/switchboard/ctc/feat/baseline_args.py [hunk 21]
```diff
@@ -237,6 +237,54 @@ def get_returnn_config(
         # network["source"] = specaug_layer_jingjing(in_layer=["features"])
         pass
 
+    if audio_perturbation and recognition:
+        feature_net = copy.deepcopy(feature_net)
+        audio_perturb_args = extra_args.get("audio_perturb_args", {})
+        source_layer = "data"
+        if "preemphasis" in audio_perturb_args:
```
Comment (vieting) [inline users/vieting/experiments/switchboard/ctc/feat/baseline_args.py:244]:
```suggestion
        if "preemphasis" in audio_perturb_args:
            # preemphasis in training is done in perturbation pre-processing, add to network for recognition
```

File: users/vieting/experiments/switchboard/ctc/feat/baseline_args.py [hunk 22]
```diff
@@ -237,6 +237,54 @@ def get_returnn_config(
...
+            for layer in feature_net["subnetwork"]:
+                if feature_net["subnetwork"][layer].get("from", None) == source_layer:
+                    feature_net["subnetwork"][layer]["from"] = "preemphasis"
+            source_layer = "preemphasis"
+            alpha = audio_perturb_args["preemphasis"].get("default")
+            feature_net["subnetwork"]["preemphasis"] = PreemphasisNetwork(alpha=alpha).get_as_subnetwork()
```
Comment (vieting) [inline users/vieting/experiments/switchboard/ctc/feat/baseline_args.py:250]:
```suggestion
            feature_net["subnetwork"]["preemphasis"] = PreemphasisNetwork(alpha=alpha).get_as_subnetwork(
                source=source_layer
            )
```

File: users/vieting/experiments/switchboard/ctc/feat/baseline_args.py [hunk 23]
```diff
@@ -237,6 +237,54 @@ def get_returnn_config(
         # network["source"] = specaug_layer_jingjing(in_layer=["features"])
         pass
 
+    if audio_perturbation and recognition:
+        feature_net = copy.deepcopy(feature_net)
+        audio_perturb_args = extra_args.get("audio_perturb_args", {})
```
Comment (vieting) [inline users/vieting/experiments/switchboard/ctc/feat/baseline_args.py:242]:
```suggestion
        audio_perturb_args = extra_args.get("audio_perturb_args", {})
        assert not ("preemphasis" in audio_perturb_args and "codecs" in audio_perturb_args), (
            "Not implemented yet, need to think about the order to apply"
        )
```

File: users/vieting/experiments/switchboard/ctc/feat/baseline_args.py [hunk 24]
```diff
@@ -237,6 +237,54 @@ def get_returnn_config(
...
+        source_layer = "data"
+        if "preemphasis" in audio_perturb_args:
+            for layer in feature_net["subnetwork"]:
+                if feature_net["subnetwork"][layer].get("from", None) == source_layer:
+                    feature_net["subnetwork"][layer]["from"] = "preemphasis"
+            source_layer = "preemphasis"
+            alpha = audio_perturb_args["preemphasis"].get("default")
+            feature_net["subnetwork"]["preemphasis"] = PreemphasisNetwork(alpha=alpha).get_as_subnetwork()
+        if "codecs" in audio_perturb_args:
```
Comment (vieting) [inline users/vieting/experiments/switchboard/ctc/feat/baseline_args.py:251]:
```suggestion
        if "codecs" in audio_perturb_args:
            # codec in training is applied in perturbation pre-processing, add to network for recognition
```

File: users/vieting/experiments/switchboard/ctc/feat/baseline_args.py [hunk 25]
```diff
@@ -237,6 +237,54 @@ def get_returnn_config(
...
+            for layer in feature_net["subnetwork"]:
+                if feature_net["subnetwork"][layer].get("from", None) == source_layer:
+                    feature_net["subnetwork"][layer]["from"] = "codec"
+            source_layer = "codec"
+            feature_net["subnetwork"]["codec"] = {
+                "class": "subnetwork",
+                "from": ["data"],
```
Comment (vieting) [inline users/vieting/experiments/switchboard/ctc/feat/baseline_args.py:258]:
```suggestion
                "from": source_layer,
```

**Review by vieting — CHANGES_REQUESTED** [review]
With the requested changes, I think we can merge the PR.

**Comment by Max-Ryujin** [comment]
All changes are applied. PR should be ready now.
